packages/xo-web/src/common/xo/index.js

```diff
@@ -1806,7 +1806,32 @@ export const editVm = async (vm, props) => {
   }
   await _call('vm.set', { ...props, id: resolveId(vm) })
     .catch(err => {
-      error(_('setVmFailed', { vm: renderXoItemFromId(resolveId(vm)) }), err.message)
+      // error code 13 is vmBadPowerState "VM state is xxx but should be halted"
+      if ((err.code === 13 && err.data.actual === 'running') || err.message === 'Cannot change memory on running VM') {
+        return chooseAction({
+          body: <p>{_('vmEditAndRestartModalMessage')}</p>,
+          buttons: [
+            { label: _('rebootVmLabel'), value: false, btnStyle: 'success' },
+            { label: _('forceRebootVmLabel'), value: true, btnStyle: 'danger' },
+          ],
+          icon: 'vm-reboot',
+          title: _('vmEditAndRestartModalTitle'),
+        })
+          .then(force => {
+            return _call('vm.setAndRestart', { ...props, id: resolveId(vm), force })
+              .then(noop)
+              .catch(err => {
+                error(_('setAndRestartVmFailed', { vm: renderXoItemFromId(resolveId(vm)) }), err.message)
+              })
+          })
+          .catch(err => {
+            if (err !== undefined) {
+              throw err
+            }
+          })
+      } else {
+        error(_('setVmFailed', { vm: renderXoItemFromId(resolveId(vm)) }), err.message)
+      }
```

Comment:
If you think it can make it easier to read and maintain, you can switch the `.then()/.catch()` to `try/await/catch`.